### ISSUE TITLE
local $^D gives noise

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7539]
Migrated from [rt.perl.org#31971](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=31971) (status was 'resolved')

Searchable as RT31971$

**Comment by p5pRT** [comment]
### From nick@llama.ing-simmons.net

#### Created by nick@llama.ing-simmons.net



The use of local $^D still works to
locally turn on some debug but
you get a lot of "stuff" as well&#8203;:

\#\!perl

sub foo \{\}

foo\(\);
\{

 local $^D = 8;
 foo\(\);
\}
\_\_END\_\_

perl /tmp/ex
 Debugging flag values&#8203;: \(see also \-d\)
&nbsp;  p  Tokenizing and parsing \(with v\, displays parse stack\)
&nbsp;  s  Stack snapshots\. with v\, displays all stacks
&nbsp;  l  Context \(loop\) stack processing
&nbsp;  t  Trace execution
&nbsp;  o  Method and overloading resolution
&nbsp;  c  String/numeric conversions
&nbsp;  P  Print profiling info\, preprocessor command for \-P\, source file input state
&nbsp;  m  Memory allocation
&nbsp;  f  Format processing
&nbsp;  r  Regular expression parsing and execution
&nbsp;  x  Syntax tree dump
&nbsp;  u  Tainting checks \(Obsolete\, previously used for LEAKTEST\)
&nbsp;  H  Hash dump \-\- usurps values\(\)
&nbsp;  X  Scratchpad allocation
&nbsp;  D  Cleaning up
&nbsp;  S  Thread synchronization
&nbsp;  T  Tokenising
&nbsp;  R  Include reference counts of dumped variables \(eg when using \-Ds\)
&nbsp;  J  Do not s\,t\,P\-debug \(Jump over\) opcodes within package DB
&nbsp;  v  Verbose&#8203;: use in conjunction with other flags
&nbsp;  C  Copy On Write
&nbsp;  A  Consistency checks on internal structures
&nbsp;  q  quiet \- currently only suppressed the 'EXECUTING' message
\(/tmp/ex&#8203;:8\)     nextstate
\(/tmp/ex&#8203;:9\)     pushmark
\(/tmp/ex&#8203;:9\)     gv\(main&#8203;::foo\)
\(/tmp/ex&#8203;:9\)     entersub
\(/tmp/ex&#8203;:9\)     nextstate
\(/tmp/ex&#8203;:3\)     leavesub
\(/tmp/ex&#8203;:9\)     leaveloop

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.5:

Configured by nick at Tue Jul 20 09:06:09 BST 2004.

Summary of my perl5 (revision 5 version 8 subversion 5) configuration:
  Platform:
    osname=linux, osvers=2.6.5-7.95-default, archname=i686-linux-multi
    uname='linux llama 2.6.5-7.95-default #1 thu jul 1 15:23:45 utc 2004 i686 athlon i386 gnulinux '
    config_args='-der -O -Doptimize=-O2 -g -Dusemymalloc=n -Dprefix=/usr/local/perl -Dusevfork=false -Dusemultiplicity=y'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-DDEBUGGING -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-DDEBUGGING -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion='', gccversion='3.3.3 (SuSE Linux)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.3.3'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    MAINT23139


@INC for perl v5.8.5:
    /usr/local/perl/lib/5.8.5/i686-linux-multi
    /usr/local/perl/lib/5.8.5
    /usr/local/perl/lib/site_perl/5.8.5/i686-linux-multi
    /usr/local/perl/lib/site_perl/5.8.5
    /usr/local/perl/lib/site_perl/5.8.4/i686-linux-multi
    /usr/local/perl/lib/site_perl/5.8.4
    /usr/local/perl/lib/site_perl
    .


Environment for perl v5.8.5:
    HOME=/home/nick
    LANG=en_GB.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=.:/home/nick/bin:.:/home/nick/bin:/home/nick/bin:/usr/local/bin:/usr/bin:/usr/X11R6/bin:/bin:/usr/games:/opt/gnome/bin:/opt/kde3/bin:/usr/lib/java/bin:/usr/local/bin:/usr/sbin:/sbin:/usr/local/bin:/usr/sbin:/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From nick@ing-simmons.net

Nick @&#8203; Llama \. Ing\-Simmons \. Net \<perl5\-porters@&#8203;perl\.org> writes&#8203;:
>\# New Ticket Created by  nick@&#8203;llama\.ing\-simmons\.net 
>\# Please include the string&#8203;:  \[perl \#31971\]
>\# in the subject line of all future correspondence about this issue\. 
>\# \<URL&#8203;: http&#8203;://rt\.perl\.org&#8203;:80/rt3/Ticket/Display\.html?id=31971 >
>
>
>
>This is a bug report for perl from nick@&#8203;llama\.ing\-simmons\.net\,
>generated with the help of perlbug 1\.35 running under perl v5\.8\.5\.
>
>
>\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>\[Please enter your report here\]
>
>The use of local $^D still works to
>locally turn on some debug but
>you get a lot of "stuff" as well&#8203;:
>
>\#\!perl
>
>sub foo \{\}
>
>foo\(\);
>\{
>
> local $^D = 8;
> foo\(\);
>\}
>\_\_END\_\_
>
>perl /tmp/ex
> Debugging flag values&#8203;: \(see also \-d\)
>  p  Tokenizing and parsing \(with v\, displays parse stack\)
>  s  Stack snapshots\. with v\, displays all stacks
>  l  Context \(loop\) stack processing
>  t  Trace execution
>  o  Method and overloading resolution
>  c  String/numeric conversions
>  P  Print profiling info\, preprocessor command for \-P\, source file input state
>  m  Memory allocation
>  f  Format processing
>  r  Regular expression parsing and execution
>  x  Syntax tree dump
>  u  Tainting checks \(Obsolete\, previously used for LEAKTEST\)
>  H  Hash dump \-\- usurps values\(\)
>  X  Scratchpad allocation
>  D  Cleaning up
>  S  Thread synchronization
>  T  Tokenising
>  R  Include reference counts of dumped variables \(eg when using \-Ds\)
>  J  Do not s\,t\,P\-debug \(Jump over\) opcodes within package DB
>  v  Verbose&#8203;: use in conjunction with other flags
>  C  Copy On Write
>  A  Consistency checks on internal structures
>  q  quiet \- currently only suppressed the 'EXECUTING' message

This seems to be&#8203;:

Change 19870 on 2003/06/28 by rgs@&#8203;rgs\-home

&nbsp;    Two debugging patches\.
&nbsp;    The first allows to hold symbolic switches in $^D
&nbsp;    and more generally fixes assignment to $^D\. The
&nbsp;    second one improves the information given by \-Dl\.
&nbsp;    
&nbsp;    Subject&#8203;: \[PATCH\] allow $^D = "flags"
&nbsp;    From&#8203;: Dave Mitchell \<davem@&#8203;fdgroup\.com>
&nbsp;    Date&#8203;: Fri\, 27 Jun 2003 22&#8203;:26&#8203;:24 \+0100
&nbsp;    Message\-ID&#8203;: \<20030627212624\.GB12887@&#8203;fdgroup\.com>

The problem is that setting 

$^D = "";

That occurs as part of the MAGIC stuff triggers 
the usage message as null string doesn't start 
with either alpha or digit\.

As giving the list on \-D'' seems to be part of new functionality
it seems mg\.c should avoid calling get\_debug\_opts\(\) in that 
case\.

Perhaps&#8203;:

&nbsp;    case '\\004'&#8203;:    /\* ^D \*/
\#ifdef DEBUGGING
&nbsp;    s = SvPV\_nolen\(sv\);
&nbsp;    PL\_debug = \(\(\*s\) ? get\_debug\_opts\(&s\) : 0\)|DEBUG\_TOP\_FLAG;
&nbsp;    DEBUG\_x\(dump\_all\(\)\);
\#else
&nbsp;    PL\_debug = \(SvIOK\(sv\) ? SvIVX\(sv\) : sv\_2iv\(sv\)\) | DEBUG\_TOP\_FLAG;
\#endif
&nbsp;    break;




>\(/tmp/ex&#8203;:8\)     nextstate
>\(/tmp/ex&#8203;:9\)     pushmark
>\(/tmp/ex&#8203;:9\)     gv\(main&#8203;::foo\)
>\(/tmp/ex&#8203;:9\)     entersub
>\(/tmp/ex&#8203;:9\)     nextstate
>\(/tmp/ex&#8203;:3\)     leavesub
>\(/tmp/ex&#8203;:9\)     leaveloop
>
>
>
>
>
>\[Please do not change anything below this line\]
>\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>\-\-\-
>Flags&#8203;:
>    category=core
>    severity=low
>\-\-\-
>Site configuration information for perl v5\.8\.5&#8203;:
>
>Configured by nick at Tue Jul 20 09&#8203;:06&#8203;:09 BST 2004\.
>
>Summary of my perl5 \(revision 5 version 8 subversion 5\) configuration&#8203;:
>  Platform&#8203;:
>    osname=linux\, osvers=2\.6\.5\-7\.95\-default\, archname=i686\-linux\-multi
>    uname='linux llama 2\.6\.5\-7\.95\-default \#1 thu jul 1 15&#8203;:23&#8203;:45 utc 2004 i686 athlon i386 gnulinux '
>    config\_args='\-der \-O \-Doptimize=\-O2 \-g \-Dusemymalloc=n \-Dprefix=/usr/local/perl \-Dusevfork=false \-Dusemultiplicity=y'
>    hint=recommended\, useposix=true\, d\_sigaction=define
>    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=define
>    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
>    use64bitint=undef use64bitall=undef uselongdouble=undef
>    usemymalloc=n\, bincompat5005=undef
>  Compiler&#8203;:
>    cc='cc'\, ccflags ='\-DDEBUGGING \-fno\-strict\-aliasing \-pipe \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
>    optimize='\-O2 \-g'\,
>    cppflags='\-DDEBUGGING \-fno\-strict\-aliasing \-pipe \-I/usr/local/include'
>    ccversion=''\, gccversion='3\.3\.3 \(SuSE Linux\)'\, gccosandvers=''
>    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
>    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
>    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
>    alignbytes=4\, prototype=define
>  Linker and Libraries&#8203;:
>    ld='cc'\, ldflags =' \-L/usr/local/lib'
>    libpth=/usr/local/lib /lib /usr/lib
>    libs=\-lnsl \-lgdbm \-ldb \-ldl \-lm \-lcrypt \-lutil \-lc
>    perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
>    libc=\, so=so\, useshrplib=false\, libperl=libperl\.a
>    gnulibc\_version='2\.3\.3'
>  Dynamic Linking&#8203;:
>    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
>    cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'
>
>Locally applied patches&#8203;:
>    MAINT23139
>
>\-\-\-
>@&#8203;INC for perl v5\.8\.5&#8203;:
>    /usr/local/perl/lib/5\.8\.5/i686\-linux\-multi
>    /usr/local/perl/lib/5\.8\.5
>    /usr/local/perl/lib/site\_perl/5\.8\.5/i686\-linux\-multi
>    /usr/local/perl/lib/site\_perl/5\.8\.5
>    /usr/local/perl/lib/site\_perl/5\.8\.4/i686\-linux\-multi
>    /usr/local/perl/lib/site\_perl/5\.8\.4
>    /usr/local/perl/lib/site\_perl
>    \.
>
>\-\-\-
>Environment for perl v5\.8\.5&#8203;:
>    HOME=/home/nick
>    LANG=en\_GB\.UTF\-8
>    LANGUAGE \(unset\)
>    LD\_LIBRARY\_PATH \(unset\)
>    LOGDIR \(unset\)
>    PATH=\.&#8203;:/home/nick/bin&#8203;:\.&#8203;:/home/nick/bin&#8203;:/home/nick/bin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/usr/X11R6/bin&#8203;:/bin&#8203;:/usr/games&#8203;:/opt/gnome/bin&#8203;:/opt/kde3/bin&#8203;:/usr/lib/java/bin&#8203;:/usr/local/bin&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/local/bin&#8203;:/usr/sbin&#8203;:/sbin
>    PERL\_BADLANG \(unset\)
>    SHELL=/bin/bash

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @rgs

nick@&#8203;llama\.ing\-simmons\.net \(via RT\) wrote&#8203;:
> The use of local $^D still works to
> locally turn on some debug but
> you get a lot of "stuff" as well&#8203;:

Fixed\, the brute\-force way\, by change \#23365 to blead\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From nick@ing-simmons.net

Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;mandrakesoft\.com> writes&#8203;:
>nick@&#8203;llama\.ing\-simmons\.net \(via RT\) wrote&#8203;:
>> The use of local $^D still works to
>> locally turn on some debug but
>> you get a lot of "stuff" as well&#8203;:
>
>Fixed\, the brute\-force way\, by change \#23365 to blead\.

Looks good\, that kind of change was what I was planning to do as well\.
